core/src/main/java/org/elasticsearch/plugins/EnginePlugin.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.engine.EngineFactory;
+
+import java.util.Optional;
+
+/**
+ * A plugin that provides alternative engine implementations.
+ */
+public interface EnginePlugin {
+
+    /**
+     * When an index is created this method is invoked for each engine plugin. Engine plugins can inspect the index settings to determine
+     * whether or not to provide an engine factory for the given index. A plugin that is not overriding the default engine should return
+     * {@link Optional#empty()}. If multiple plugins return an engine factory for a given index the index will not be created and an
+     * {@link IllegalStateException} will be thrown during index creation.
+     *
+     * @return an optional engine factory
+     */
+    Optional<EngineFactory> getMaybeEngineFactory(IndexSettings indexSettings);
```

Comment:
Why the use of Optional? In other places we have just used `null` to indicate there is no engine implementation being added. Also, can I don't think we should use the `Maybe` style of naming.  Again, in the pull based plugin APIs we have not done this anywhere else, and I don't think it really adds anything.

Reply:
+1 to null and rename to `getEngineFactory`

Reply:
I prefer the use of optional, but I'm fine with renaming.

Reply:
I think we should come to an agreement on whether we should use Optional for these, and be consistent. But arbitrarily adding inconsistencies into the API will just create confusion.

Reply:
I agree on the consistency argument. We discussed `Optional` versus `null` in a separate channel that we will merge this as-is for now (with `Optional`), consider changing the handful of other places in the plugin API where we allow `null` to be returned but will have a broader discussion about this first. We can return to this change if needed.